clang-tidy fails with modernize diagnostics and standard C++ 20 or newer

clang-tidy v21.1.5 fails on macOS (Homebrew version) when modernize diagnostics are set and standard is more recent than C++17.

Let's consider this `hello.cpp`:

```c++
#include <cstddef>
#include <iostream>

int main([[maybe_unused]] int argc, [[maybe_unused]] char** argv)
{
    std::cout << "Hello world!" << std::endl;

    return EXIT_SUCCESS;
}
```

With C++ 17 it works:

```shell
➜ clang-tidy hello.cpp -checks="-*,modernize-*" -- -std=c++17 -x c++    

11204 warnings generated.
/Users/sgilles/Codes/Sandbox/hello.cpp:3:5: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
    3 | int main(int argc, char** argv)
      | ~~~ ^                          
      | auto                            -> int
Suppressed 11204 warnings (11203 in non-user code, 1 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```

but it fails with more recent standards:

```shell
➜ clang-tidy hello.cpp -checks="-*,modernize-*" -- -std=c++20 -x c++
13289 warnings and 10 errors generated.
Error while processing /Users/sgilles/Codes/Sandbox/hello.cpp.
error: missing '(' after '__has_feature' [clang-diagnostic-error]
/Users/sgilles/Codes/Sandbox/hello.cpp:3:5: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
    3 | int main(int argc, char** argv)
      | ~~~ ^                          
      | auto                            -> int
Suppressed 13290 warnings (13288 in non-user code, 2 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler error(s).
```

This is recent (the exact same code worked with the diagnostic and C++ 23 standard few months ago) and it is related specifically to `modernize` diagnostics.

--- a/clang-tools-extra/clang-tidy/modernize/UseTrailingReturnTypeCheck.cpp
+++ b/clang-tools-extra/clang-tidy/modernize/UseTrailingReturnTypeCheck.cpp
@@ -269,10 +269,8 @@ classifyTokensBeforeFunctionName(const FunctionDecl &F, const ASTContext &Ctx,
 
       if (Info.hasMacroDefinition()) {
         const MacroInfo *MI = PP->getMacroInfo(&Info);
-        if (!MI || MI->isFunctionLike()) {
-          // Cannot handle function style macros.
+        if (!MI || MI->isFunctionLike() || MI->isBuiltinMacro())
           return std::nullopt;
-        }
       }
 
       T.setIdentifierInfo(&Info);

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -265,6 +265,11 @@ Changes in existing checks
   <clang-tidy/checks/modernize/use-std-format>` check by fixing a crash
   when an argument is part of a macro expansion.
 
+- Improved :doc:`modernize-use-trailing-return-type
+  <clang-tidy/checks/modernize/use-trailing-return-type>` check by fixing
+  spurious ``missing '(' after '__has_feature'`` errors caused by builtin
+  macros appearing in the return type of a function.
+
 - Improved :doc:`modernize-use-using
   <clang-tidy/checks/modernize/use-using>` check by avoiding the generation
   of invalid code for function types with redundant parentheses.

--- a/clang-tools-extra/test/clang-tidy/checkers/modernize/use-trailing-return-type.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/modernize/use-trailing-return-type.cpp
@@ -487,6 +487,17 @@ decltype(COMMAND_LINE_INT{}) h21();
 // CHECK-MESSAGES: :[[@LINE-1]]:30: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
 // CHECK-FIXES: auto h21() -> decltype(COMMAND_LINE_INT{});
 
+// Builtin macros like __has_feature are registered as object-like macros but
+// require function-like syntax when expanded. Ensure the check does not cause
+// spurious "missing '(' after '__has_feature'" errors when they appear in the
+// raw lex span of a function return type.
+const decltype(__has_feature(cxx_constexpr)) h22();
+// CHECK-MESSAGES: :[[@LINE-1]]:46: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
+// CHECK-FIXES: const decltype(__has_feature(cxx_constexpr)) h22();
+const decltype(__has_builtin(__builtin_expect)) h23();
+// CHECK-MESSAGES: :[[@LINE-1]]:49: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
+// CHECK-FIXES: const decltype(__has_builtin(__builtin_expect)) h23();
+
 //
 // Name collisions
 //

